### PR TITLE
Add PIMO

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
## 📝 Description

- Adds the PIMO metric created by @jpcbertoldo 

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/metrics/__init__.py [hunk 1]
```diff
@@ -35,6 +37,9 @@
     "ManualThreshold",
     "MinMax",
     "PRO",
+    "pimo",
```
Comment (samet-akcay) [inline src/anomalib/metrics/__init__.py:40]: is this relative import needed here?

File: src/anomalib/metrics/pimo/__init__.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+"""Per-Image Metrics."""
+
+# Original Code
+# https://github.com/jpcbertoldo/aupimo
+#
+# Modified
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from .enums import StatsOutliersPolicy, StatsRepeatedPolicy, ThresholdMethod
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/__init__.py:10]: Joao was mentioning that these policies could also be removed. Do you think we should remove them together with the validators?

File: src/anomalib/metrics/pimo/binary_classification_curve.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        anomaly_maps (torch.Tensor): Anomaly score maps of shape (N, H, W)
+        masks (torch.Tensor): Binary ground truth masks of shape (N, H, W)
+        threshs_choice (str, optional): Sequence of thresholds to use. Defaults to THRESH_SEQUENCE_MINMAX_LINSPACE.
+        #
+        # `threshs_choice`-dependent arguments
+        #
+        # THRESH_SEQUENCE_GIVEN
+        threshs_given (torch.Tensor, optional): Sequence of thresholds to use.
+        #
+        # THRESH_SEQUENCE_MINMAX_LINSPACE
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/binary_classification_curve.py:185]: is this commented block sphinx safe?

File: src/anomalib/metrics/pimo/dataclasses.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            _validate.is_thresh_bounds((self.thresh_lower_bound, self.thresh_upper_bound))
+
+        except (TypeError, ValueError) as ex:
+            msg = f"Invalid inputs for {self.__class__.__name__} object. Cause: {ex}."
+            raise TypeError(msg) from ex
+
+    @classmethod
+    def from_pimoresult(
+        cls: type["AUPIMOResult"],
+        pimoresult: PIMOResult,
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/dataclasses.py:184]:
```suggestion
        pimo_result: PIMOResult,
```

File: src/anomalib/metrics/pimo/dataclasses.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            _validate.is_rates(self.aupimos, nan_allowed=True)  # validate is_aupimos
+
+            _validate.is_thresh_bounds((self.thresh_lower_bound, self.thresh_upper_bound))
+
+        except (TypeError, ValueError) as ex:
+            msg = f"Invalid inputs for {self.__class__.__name__} object. Cause: {ex}."
+            raise TypeError(msg) from ex
+
+    @classmethod
+    def from_pimoresult(
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/dataclasses.py:182]:
```suggestion
    def from_pimo_result(
```

File: src/anomalib/metrics/pimo/functional.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+from .enums import ThresholdMethod
+from .utils import images_classes_from_masks
+
+logger = logging.getLogger(__name__)
+
+
+def pimo_curves(
+    anomaly_maps: torch.Tensor,
+    masks: torch.Tensor,
+    num_threshs: int,
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/functional.py:34]: can we rename `num_threshs` to `num_thresholds`?

File: src/anomalib/metrics/pimo/pimo.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+    @property
+    def image_classes(self) -> torch.Tensor:
+        """Image classes (0: normal, 1: anomalous)."""
+        return functional.images_classes_from_masks(self.masks)
+
+    def __init__(
+        self,
+        num_threshs: int,
+    ) -> None:
```
Comment (samet-akcay) [inline src/anomalib/metrics/pimo/pimo.py:112]:
```suggestion
    def __init__(self, num_thresholds: int) -> None:
```

File: tests/unit/metrics/pimo/test_binclf_curve.py
Comment (samet-akcay) [inline tests/unit/metrics/pimo/test_binclf_curve.py:1]: now that we renamed the `binclf`, should we rename this module as well?

**Review by samet-akcay — COMMENTED** [review]
thanks, some minor comments only.

overall, can we rename all the `thresh` to `threshold` and `threshs` to `thresholds`

**Review by samet-akcay — APPROVED** [review]

**Comment by jpcbertoldo** [comment]
Yay!

**Pull request merged**